api/types/database_test.go

```diff
@@ -564,6 +566,22 @@ func TestDynamoDBConfig(t *testing.T) {
 				},
 			},
 		},
+		{
+			desc:       "account and region and assume role is correct",
+			region:     "us-west-1",
+			account:    "123456789012",
+			roleARN:    "arn:aws:iam::123456789012:role/DBDiscoverer",
+			externalID: "externalid123",
+			wantSpec: DatabaseSpecV3{
+				URI: "aws://dynamodb.us-west-1.amazonaws.com",
+				AWS: AWS{
+					Region:        "us-west-1",
+					AccountID:     "123456789012",
+					AssumeRoleARN: "arn:aws:iam::123456789012:role/DBDiscoverer",
```

Comment:
Curious why `AssumeRoleARN` is required in this case? I could be wrong but I feel `tsh db connect --db-user arn:aws:iam::<another-account>:role/dynamodb-role my-dynamodb` is legit with `ExternalID` configured?

Reply:
Good point. We have 3 protocols (cassandra, dynamodb, redshift serverless) that assume a role for authentication, and the role is built from `--db-user`.

Of those 3, only dynamodb is passing `external_id` when it assumes the role.

I discussed with @smallinsky that for these protocols we might have two roles to assume but only one external id - so which role should pair with the external ID?

We decided it would be be best to pair external ID with the configured `assume_role_arn`, so that cross-account access delegation only needs to be setup for one role rather than multiple.

The problem with that change is DynamoDB, which is already supporting `external_id`...

I suppose I could remove this check for cassandra/redshiftserverless/dynamodb specifically and just pass `external_id` with both assume role calls, since it won't have any effect for roles that are already in the same account as the caller. That would prevent breaking existing config for anyone using external ID w/ DynamoDB, and make the behavior consistent for all 3 protocols. What do you guys think? Any issues with doing that?

Reply:
An alternative solution is to check the database's configured `AccountID` and only pass external id when assuming a configured `assume_role_arn` if the account ID doesn't match the role account ID, and likewise for assuming a role from `--db-user`. That might be a cleaner solution.

Reply:
hmm. chained assume role? that's interesting. Would it make sense to have two external IDs for each? not sure if this is more confusing but something like this:

host role -> assume "discovery role" with `aws.AssumeRole.ARN` and `aws.AssumeRole.ExternalID` -> assume db user role with `--db-user` and `aws.ExternalID`

Reply:
> I suppose I could remove this check for cassandra/redshiftserverless/dynamodb specifically and just pass external_id with both assume role calls, since it won't have any effect for roles that are already in the same account as the caller. That would prevent breaking existing config for anyone using external ID w/ DynamoDB, and make the behavior consistent for all 3 protocols. What do you guys think? Any issues with doing that?

Or, only use external_id for "discovery role" if it's set, and if "discovery role" not set, use it for the "db user role"?

Reply:
> host role -> assume "discovery role" with aws.AssumeRole.ARN and aws.AssumeRole.ExternalID -> assume db user role with --db-user and aws.ExternalID

> only use external_id for "discovery role" if it's set, and if "discovery role" not set, use it for the "db user role"?

I think it might be too confusing to have two configurable external IDs. I prefer the second suggestion. For config validation, that means I'll have a special exception that dynamodb/redshift serverless/cassandra will not require `assume_role_arn` with `external_id`.

Reply:
And to add onto this, I think if `assume_role_arn` is given, no matter the db protocol, we should require it matches `account_id` (or if `account_id` is blank, set it to match `assume_role_arn`).

Reasoning is as follows:
 
Technically for fetching an auth token or signing requests, `--db-user` could provide a role in the same account as the database. However, for metadata fetching and IAM configuration, the agent will only know about `assume_role_arn`, and those operations will fail if the database is in a different account than the role that the agent assumes.

Example:
```yaml
db_service:
  enabled: true
  databases:
    - name: "ddb"
      protocol: "dynamodb"
      static_labels:
        foo: "bar"
      aws:
        region: "us-west-1"
        account_id: "123456789012"
        assume_role_arn: "arn:aws:iam::278576220453:role/role-in-external-account"
        external_id: "1234"
```

```
# works for proxying requests, but agent IAM configurator is failing since it has an external `assume_role_arn`
tsh proxy db --tunnel --db-user="arn:aws:iam::123456789012:role/role-in-same-account" ddb
```

If `assume_role_arn` is instead left blank, then the agent IAM configurator will use `sts get-caller-identity` to configure itself.
If `assume_role_arn` is not blank but matches the database account ID, then agent IAM configurator will work in an external account.

Reply:
note that in the case where `--db-user` is an external account and `assume_role_arn` is blank, the agent's configurator won't be able to configure IAM across accounts because it doesn't know about potential settings for `--db-user`. This limitation means we should recommend in docs that for cross account AWS, they should use `assume_role_arn` with `external_id` instead of supplying the external role via `--db-user`.

We do not currently have IAM configurator for DynamoDB, so existing customers are having to configure IAM manually already anyway. I plan to add that as well. I don't think this will regress anything.

Reply:
I made the change for `CheckAndSetDefaults`. It only errors if `d.RequireAWSIAMRolesAsUsers()` returns false. This covers keyspaces, redshift serverless, and dynamodb.